# ML Zoomcamp 2026 — Homework 1: Introduction to Machine Learning

Dataset: 2026 Car Fuel Efficiency (`car_fuel_efficiency_2026.csv`)

In [1]:
import numpy as np
import pandas as pd

## Q1. Pandas version

In [2]:
pd.__version__

'3.0.2'

## Getting the data

Reads the local file if present, otherwise downloads it from the course repository.

In [3]:
import os

fname = 'car_fuel_efficiency_2026.csv'
url = 'https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/main/cohorts/2026/data/' + fname

df = pd.read_csv(fname if os.path.exists(fname) else url)
df.head()

,model_year,origin,fuel_type,drivetrain,num_doors,engine_displacement,num_cylinders,horsepower,vehicle_weight,acceleration,fuel_efficiency_mpg
0,2006,Europe,Gasoline,Front-wheel drive,4,2180,6,243.0,3870,NaN,31.9
1,2008,Europe,Diesel,Front-wheel drive,4,2390,6,272.0,4210,NaN,31.3
2,1996,Asia,Gasoline,Front-wheel drive,5,2320,6,267.0,4240,17.3,27.5
3,1989,Europe,Gasoline,Front-wheel drive,4,2130,6,258.0,4490,18.7,28.5
4,1994,USA,Diesel,Front-wheel drive,3,2580,7,304.0,4510,17.5,31.0


## Q2. Records count

In [4]:
len(df)

10000

## Q3. Fuel types

In [5]:
print(df['fuel_type'].value_counts())
df['fuel_type'].nunique()

fuel_type
Gasoline    6913
Diesel      2053
Hybrid      1034
Name: count, dtype: int64


3

## Q4. Missing values

In [6]:
missing = df.isnull().sum()
print(missing)
print()
print('Columns with missing values:', (missing > 0).sum())

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64

Columns with missing values: 2


## Q5. Max fuel efficiency of cars from Asia

In [7]:
df[df['origin'] == 'Asia']['fuel_efficiency_mpg'].max()

np.float64(41.2)

## Q6. Median value of horsepower

In [8]:
median_before = df['horsepower'].median()
most_frequent = df['horsepower'].mode()[0]
median_after = df['horsepower'].fillna(most_frequent).median()

print('Median before:', median_before)
print('Most frequent value:', most_frequent)
print('Median after fillna:', median_after)

Median before: 254.0
Most frequent value: 252.0
Median after fillna: 252.0


## Q7. Sum of weights

In [9]:
X = df[df['origin'] == 'Asia'][['vehicle_weight', 'model_year']].head(7).to_numpy()

XTX = X.T @ X
XTX_inv = np.linalg.inv(XTX)

y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = XTX_inv @ X.T @ y

print('w =', w)
print('Sum of w =', w.sum())

w = [0.13644777 0.2327492 ]
Sum of w = 0.36919696904925486


## Answers

| Question | Answer |
|---|---|
| Q1. Pandas version | see `pd.__version__` above |
| Q2. Records count | 10000 |
| Q3. Fuel types | 3 |
| Q4. Missing values | 2 |
| Q5. Max fuel efficiency | 41.2 |
| Q6. Median value of horsepower | Yes, it decreased |
| Q7. Sum of weights | 0.369 |